### PR TITLE
Add Detailed Sample Creation Date

### PR DESCRIPTION
**Pull request by Kashif-S**

### CODE REVIEW

**Comment by sqwci** [comment]
SonarQube analysis reported 27 issues
* ![CRITICAL][CRITICAL] 6 critical
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 19 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DetailedSampleBuilder.java#L994](https://github.com/TGAC/miso-lims/blob/c7e8a414a87fe3c9baca73f1af2f8828ef22e96a/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/DetailedSampleBuilder.java#L994): Define a constant instead of duplicating this literal "Cannot delete non-persisted builder object" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [DefaultSampleService.java#L225](https://github.com/TGAC/miso-lims/blob/c7e8a414a87fe3c9baca73f1af2f8828ef22e96a/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L225): Refactor this method to reduce its Cognitive Complexity from 28 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![CRITICAL][CRITICAL] [DefaultSampleService.java#L277](https://github.com/TGAC/miso-lims/blob/c7e8a414a87fe3c9baca73f1af2f8828ef22e96a/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L277): Refactor this method to reduce its Cognitive Complexity from 18 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![CRITICAL][CRITICAL] [EditSampleController.java#L698](https://github.com/TGAC/miso-lims/blob/c7e8a414a87fe3c9baca73f1af2f8828ef22e96a/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditSampleController.java#L698): Define a constant instead of duplicating this literal "sample" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditSampleController.java#L884](https://github.com/TGAC/miso-lims/blob/c7e8a414a87fe3c9baca73f1af2f8828ef22e96a/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditSampleController.java#L884): Define a constant instead of duplicating this literal "Samples" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditSampleController.java#L931](https://github.com/TGAC/miso-lims/blob/c7e8a414a87fe3c9baca73f1af2f8828ef22e96a/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditSampleController.java#L931): Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [DetailedSampleImpl.java#L172](https://github.com/TGAC/miso-lims/blob/c7e8a414a87fe3c9baca73f1af2f8828ef22e96a/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/DetailedSampleImpl.java#L172): Rename "groupId" which hides the field declared at line 62. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AHiddenFieldCheck)
1. ![MAJOR][MAJOR] [DefaultSampleService.java#L89](https://github.com/TGAC/miso-lims/blob/c7e8a414a87fe3c9baca73f1af2f8828ef22e96a/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L89): A "NullPointerException" could be thrown; "parent" is nullable here. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2259)
1. ![MINOR][MINOR] [DefaultSampleService.java#L419](https://github.com/TGAC/miso-lims/blob/c7e8a414a87fe3c9baca73f1af2f8828ef22e96a/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L419): Remove the declaration of thrown exception 'org.hibernate.exception.ConstraintViolationException' which is a runtime exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [DefaultSampleService.java#L509](https://github.com/TGAC/miso-lims/blob/c7e8a414a87fe3c9baca73f1af2f8828ef22e96a/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L509): Remove the declaration of thrown exception 'java.sql.SQLException', as it cannot be thrown from method's body. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by sqwci** [comment]
SonarQube analysis reported 27 issues
* ![CRITICAL][CRITICAL] 6 critical
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 19 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DetailedSampleBuilder.java#L994](https://github.com/TGAC/miso-lims/blob/fd51a74175825ac1f34e6a31fd921f43d7458000/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/DetailedSampleBuilder.java#L994): Define a constant instead of duplicating this literal "Cannot delete non-persisted builder object" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [DefaultSampleService.java#L225](https://github.com/TGAC/miso-lims/blob/fd51a74175825ac1f34e6a31fd921f43d7458000/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L225): Refactor this method to reduce its Cognitive Complexity from 28 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![CRITICAL][CRITICAL] [DefaultSampleService.java#L277](https://github.com/TGAC/miso-lims/blob/fd51a74175825ac1f34e6a31fd921f43d7458000/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L277): Refactor this method to reduce its Cognitive Complexity from 18 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![CRITICAL][CRITICAL] [EditSampleController.java#L698](https://github.com/TGAC/miso-lims/blob/fd51a74175825ac1f34e6a31fd921f43d7458000/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditSampleController.java#L698): Define a constant instead of duplicating this literal "sample" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditSampleController.java#L884](https://github.com/TGAC/miso-lims/blob/fd51a74175825ac1f34e6a31fd921f43d7458000/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditSampleController.java#L884): Define a constant instead of duplicating this literal "Samples" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditSampleController.java#L931](https://github.com/TGAC/miso-lims/blob/fd51a74175825ac1f34e6a31fd921f43d7458000/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditSampleController.java#L931): Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [DetailedSampleImpl.java#L172](https://github.com/TGAC/miso-lims/blob/fd51a74175825ac1f34e6a31fd921f43d7458000/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/DetailedSampleImpl.java#L172): Rename "groupId" which hides the field declared at line 62. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AHiddenFieldCheck)
1. ![MAJOR][MAJOR] [DefaultSampleService.java#L89](https://github.com/TGAC/miso-lims/blob/fd51a74175825ac1f34e6a31fd921f43d7458000/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L89): A "NullPointerException" could be thrown; "parent" is nullable here. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2259)
1. ![MINOR][MINOR] [DefaultSampleService.java#L419](https://github.com/TGAC/miso-lims/blob/fd51a74175825ac1f34e6a31fd921f43d7458000/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L419): Remove the declaration of thrown exception 'org.hibernate.exception.ConstraintViolationException' which is a runtime exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [DefaultSampleService.java#L509](https://github.com/TGAC/miso-lims/blob/fd51a74175825ac1f34e6a31fd921f43d7458000/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L509): Remove the declaration of thrown exception 'java.sql.SQLException', as it cannot be thrown from method's body. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/DetailedSampleImpl.java [hunk 1]
```diff
@@ -67,6 +70,9 @@
   @Column(updatable = false)
   private Long preMigrationId;
 
+  @Temporal(TemporalType.TIMESTAMP)
```
Comment (djcooke) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/DetailedSampleImpl.java]: Should be `TemporalType.DATE`

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/DetailedSample.java [hunk 2]
```diff
@@ -117,4 +118,8 @@
    */
   Optional<DetailedSample> getEffectiveGroupIdSample();
 
+  Date getCreationDate();
```
Comment (djcooke) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/DetailedSample.java]: Please Javadoc this and the setter to clarify that this is the user-specified creation date

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/DetailedSample.java [hunk 3]
```diff
@@ -118,8 +118,20 @@
    */
   Optional<DetailedSample> getEffectiveGroupIdSample();
 
+  /**
+   * Field for storing the date of sample creation. This is not the date that the sample was entered into MISO.
+   * This field is not automatically generated on sample creation and must be specified by a user.
+   * 
+   * @return Date creationDate
```
Comment (djcooke) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/DetailedSample.java]: Return type should not be included in Javadoc (`@return <description>`). Same with parameter type in the setter (`@param <name> <description>`)

**Comment by sqwci** [comment]
SonarQube analysis reported 27 issues
* ![CRITICAL][CRITICAL] 6 critical
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 19 minor

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DetailedSampleBuilder.java#L1018](https://github.com/TGAC/miso-lims/blob/217f091a7e9d439b379527cc10c481fb15040335/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/DetailedSampleBuilder.java#L1018): Define a constant instead of duplicating this literal "Cannot delete non-persisted builder object" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [DefaultSampleService.java#L225](https://github.com/TGAC/miso-lims/blob/217f091a7e9d439b379527cc10c481fb15040335/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L225): Refactor this method to reduce its Cognitive Complexity from 28 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![CRITICAL][CRITICAL] [DefaultSampleService.java#L277](https://github.com/TGAC/miso-lims/blob/217f091a7e9d439b379527cc10c481fb15040335/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L277): Refactor this method to reduce its Cognitive Complexity from 18 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![CRITICAL][CRITICAL] [EditSampleController.java#L698](https://github.com/TGAC/miso-lims/blob/217f091a7e9d439b379527cc10c481fb15040335/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditSampleController.java#L698): Define a constant instead of duplicating this literal "sample" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditSampleController.java#L888](https://github.com/TGAC/miso-lims/blob/217f091a7e9d439b379527cc10c481fb15040335/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditSampleController.java#L888): Define a constant instead of duplicating this literal "Samples" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditSampleController.java#L935](https://github.com/TGAC/miso-lims/blob/217f091a7e9d439b379527cc10c481fb15040335/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditSampleController.java#L935): Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [DetailedSampleImpl.java#L170](https://github.com/TGAC/miso-lims/blob/217f091a7e9d439b379527cc10c481fb15040335/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/DetailedSampleImpl.java#L170): Rename "groupId" which hides the field declared at line 61. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AHiddenFieldCheck)
1. ![MAJOR][MAJOR] [DefaultSampleService.java#L89](https://github.com/TGAC/miso-lims/blob/217f091a7e9d439b379527cc10c481fb15040335/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L89): A "NullPointerException" could be thrown; "parent" is nullable here. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2259)
1. ![MINOR][MINOR] [DefaultSampleService.java#L419](https://github.com/TGAC/miso-lims/blob/217f091a7e9d439b379527cc10c481fb15040335/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L419): Remove the declaration of thrown exception 'org.hibernate.exception.ConstraintViolationException' which is a runtime exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [DefaultSampleService.java#L509](https://github.com/TGAC/miso-lims/blob/217f091a7e9d439b379527cc10c481fb15040335/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L509): Remove the declaration of thrown exception 'java.sql.SQLException', as it cannot be thrown from method's body. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**